sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/TracerProviderConfiguration.java

```diff
@@ -0,0 +1,150 @@
+
+    List<SpanExporter> spanExporters = new ArrayList<>();
+    for (String name : new ArrayList<>(exporterNames)) {
+      SpanExporter exporter = SpanExporterConfiguration.configureExporter(name, config);
+      if (exporter != null) {
+        spanExporters.add(exporter);
+      }
+    }
+
+    if (!spanExporters.isEmpty()) {
+      tracerProviderBuilder.addSpanProcessor(configureSpanProcessor(config, spanExporters));
+    }
+
+    SdkTracerProvider tracerProvider = tracerProviderBuilder.build();
+    Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::shutdown));
+    return tracerProvider;
+  }
+
+  // VisibleForTesting
+  static BatchSpanProcessor configureSpanProcessor(
+      ConfigProperties config, List<SpanExporter> exporters) {
+    SpanExporter exporter = SpanExporter.composite(exporters);
+    BatchSpanProcessorBuilder builder = BatchSpanProcessor.builder(exporter);
+
+    Long scheduleDelayMillis = config.getLong("otel.bsp.schedule.delay.millis");
+    if (scheduleDelayMillis != null) {
+      builder.setScheduleDelayMillis(scheduleDelayMillis);
+    }
+
+    Integer maxQueue = config.getInt("otel.bsp.max.queue.size");
+    if (maxQueue != null) {
+      builder.setMaxQueueSize(maxQueue);
+    }
+
+    Integer maxExportBatch = config.getInt("otel.bsp.max.export.batch.size");
+    if (maxExportBatch != null) {
+      builder.setMaxExportBatchSize(maxExportBatch);
+    }
+
+    Integer timeout = config.getInt("otel.bsp.export.timeout.millis");
+    if (timeout != null) {
+      builder.setExporterTimeoutMillis(timeout);
+    }
+
+    return builder.build();
+  }
+
+  // Visible for testing
+  static TraceConfig configureTraceConfig(ConfigProperties config) {
+    TraceConfigBuilder builder = TraceConfig.getDefault().toBuilder();
+
+    String sampler = config.getString("otel.trace.sampler");
+    if (sampler != null) {
+      builder.setSampler(configureSampler(sampler, config));
+    }
+
+    Integer maxAttrs = config.getInt("otel.span.attribute.count.limit");
+    if (maxAttrs != null) {
+      builder.setMaxNumberOfAttributes(maxAttrs);
+    }
+
+    Integer maxEvents = config.getInt("otel.span.event.count.limit");
+    if (maxEvents != null) {
+      builder.setMaxNumberOfEvents(maxEvents);
+    }
+
+    Integer maxLinks = config.getInt("otel.span.link.count.limit");
+    if (maxLinks != null) {
+      builder.setMaxNumberOfLinks(maxLinks);
+    }
+
+    return builder.build();
+  }
+
+  // Visible for testing
+  static Sampler configureSampler(String sampler, ConfigProperties config) {
+    switch (sampler) {
+      case "always_on":
+        return Sampler.alwaysOn();
+      case "always_off":
+        return Sampler.alwaysOff();
+      case "traceidratio":
+        {
+          Double ratio = config.getDouble("otel.trace.sampler.arg");
+          if (ratio == null) {
+            throw new ConfigurationException(
+                "otel.trace.sampler=traceidratio but otel.trace.sampler.arg is not provided. "
```

Comment:
stuff like this is a good argument for extracting the keys into constants, but that can be done as a separate step later, for sure.